File: web/src/components/leaveByDisplay.ts
import type { LeaveByStatus } from "@/api/types"

/** Agenda leave-by copy helpers (estimate only — never live traffic / ETA). */

export const LEAVE_BY_PENDING_LABEL = "Estimating leave-by…"

export function formatLeaveByTime(iso: string): string {
  const date = new Date(iso)
  if (Number.isNaN(date.getTime())) {
    return iso
  }
  return date.toLocaleTimeString(undefined, {
    hour: "numeric",
    minute: "2-digit",
  })
}

/** e.g. "Leave by ~3:40 PM · estimate" */
export function formatLeaveByEstimateLine(leaveByAtIso: string): string {
  return `Leave by ~${formatLeaveByTime(leaveByAtIso)} · estimate`
}

/** Short human reason when leaveByStatus is UNAVAILABLE. */
export function leaveByUnavailableLabel(reason: string | null | undefined): string {
  switch (reason) {
    case "NO_ORIGIN":
      return "No leave-from place yet"
    case "NO_DESTINATION":
      return "Add a location to estimate leave-by"
    case "GEOCODE_FAILED":
      return "Couldn't locate the destination"
    default:
      return "Leave-by estimate unavailable"
  }
}

export function agendaLeaveByLine(item: {
  leaveByStatus: LeaveByStatus
  leaveByAt: string | null
  leaveByReason: string | null
}): string {
  if (item.leaveByStatus === "PENDING") {
    return LEAVE_BY_PENDING_LABEL
  }
  if (item.leaveByStatus === "OK" && item.leaveByAt) {
    return formatLeaveByEstimateLine(item.leaveByAt)
  }
  return leaveByUnavailableLabel(item.leaveByReason)
}

File: mobile/iosApp/iosApp/LeaveByDisplay.swift
import Foundation

/// Agenda leave-by copy helpers (estimate only — never live traffic / ETA).
enum LeaveByDisplay {
    private static let timeFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.timeStyle = .short
        formatter.dateStyle = .none
        return formatter
    }()

    private static let isoWithFractional: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter
    }()

    private static let isoPlain: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter
    }()

    static func formatLeaveByTime(iso: String) -> String {
        let trimmed = iso.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return iso }
        guard let date = isoWithFractional.date(from: trimmed) ?? isoPlain.date(from: trimmed) else {
            return iso
        }
        return timeFormatter.string(from: date)
    }

    static let pendingLabel = "Estimating leave-by…"

    /// e.g. "Leave by ~3:40 PM · estimate"
    static func formatLeaveByEstimateLine(leaveByAtIso: String) -> String {
        "Leave by ~\(formatLeaveByTime(iso: leaveByAtIso)) · estimate"
    }

    static func leaveByUnavailableLabel(reason: String?) -> String {
        switch reason {
        case "NO_ORIGIN":
            return "No leave-from place yet"
        case "NO_DESTINATION":
            return "Add a location to estimate leave-by"
        case "GEOCODE_FAILED":
            return "Couldn't locate the destination"
        default:
            return "Leave-by estimate unavailable"
        }
    }

    static func leaveByAgendaLine(
        leaveByStatus: String,
        leaveByAt: String?,
        leaveByReason: String?
    ) -> String {
        if leaveByStatus == "PENDING" {
            return pendingLabel
        }
        if leaveByStatus == "OK", let leaveByAt, !leaveByAt.isEmpty {
            return formatLeaveByEstimateLine(leaveByAtIso: leaveByAt)
        }
        return leaveByUnavailableLabel(reason: leaveByReason)
    }
}
